requestmanager/reconciledloader/reconciledloader.go

```diff
@@ -0,0 +1,118 @@
+/* Package reconciledloader implements a block loader that can load from two different sources:
+- a local store
+- a series of remote responses for a given graphsync selector query
+
+It verifies the sequence of remote responses matches the sequence
+of loads called from a local selector traversal.
+
+The reconciled loader also tracks whether or not there is a remote request in progress.
+
+When there is no request in progress, it loads from the local store only.
+
+When there is a request in progress, waits for remote responses before loading, and only calls
+upon the local store for duplicate blocks and when traversing paths the remote was missing.
+
+The reconciled loader assumes:
+1. A single thread is calling AsyncLoad to load blocks
+2. When a request is online, a seperate thread may call IngestResponse
+3. Either thread may call SetRemoteState or Cleanup
+4. The remote sends metadata for all blocks it traverses in the query (per GraphSync protocol spec) - whether or not
+the actual block is sent.
+*/
+package reconciledloader
+
+import (
+	"context"
+	"errors"
+	"sync"
+
+	"github.com/ipfs/go-graphsync"
+	"github.com/ipfs/go-graphsync/requestmanager/reconciledloader/traversalrecord"
+	"github.com/ipfs/go-graphsync/requestmanager/types"
+	logging "github.com/ipfs/go-log/v2"
+	"github.com/ipld/go-ipld-prime/datamodel"
+	"github.com/ipld/go-ipld-prime/linking"
+)
+
+var log = logging.Logger("gs-reconciledlaoder")
+
+type settableWriter interface {
+	SetBytes([]byte) error
+}
+
+type byteReader interface {
+	Bytes() []byte
+}
+
+type loadAttempt struct {
+	link        datamodel.Link
+	linkContext linking.LinkContext
+	successful  bool
+}
+
+func (lr loadAttempt) empty() bool {
+	return lr.link == nil
+}
+
+// ReconciledLoader is an instance of the reconciled loader
+type ReconciledLoader struct {
+	requestID             graphsync.RequestID
+	lsys                  *linking.LinkSystem
+	mostRecentOfflineLoad loadAttempt
+	traversalRecord       *traversalrecord.TraversalRecord
+	pathTracker           pathTracker
+
+	lock        *sync.Mutex
+	signal      *sync.Cond
+	open        bool
+	verifier    *traversalrecord.Verifier
+	remoteQueue remoteQueue
+}
+
+// NewReconciledLoader returns a new reconciled loader for the given requestID & localStore
+func NewReconciledLoader(requestID graphsync.RequestID, localStore *linking.LinkSystem) *ReconciledLoader {
+	lock := &sync.Mutex{}
+	traversalRecord := traversalrecord.NewTraversalRecord()
+	return &ReconciledLoader{
+		requestID:       requestID,
+		lsys:            localStore,
+		lock:            lock,
+		signal:          sync.NewCond(lock),
+		traversalRecord: traversalRecord,
+		verifier:        traversalrecord.NewVerifier(traversalRecord),
+	}
+}
+
+// SetRemoteState records whether or not the request is online
+func (rl *ReconciledLoader) SetRemoteState(online bool) {
```

Comment:
Could we rename this so it's more natural with a boolean? (or use an enum for the values, but there's really only two). Maybe `SetRemoteOnline()`? Reading `SetRemoteState(false)` is a bit mysterious.

Reply:
Yea I agree.